### PR TITLE
Mark notifications with the source qube's colour

### PR DESCRIPTION
**Pull request by oshogbo**
This is what I got:

image

### CODE REVIEW

**Comment by marmarek** [comment]
I like it, @marmarta what do you think?

**Comment by marmarek** [comment]
PipelineRetry

**Comment by marmarta** [comment]
I generally like it, but I have questions: how does it look when the icon has some transparency? I wonder if this remains readable with colorful icons - I don't see that many colorful icons in notifications in general, though.

**Comment by marmarek** [comment]
It doesn't appear to work :(
First test: https://openqa.qubes-os.org/tests/192868#step/startup/9, but here nm-applet uses `image-path`, not `image-data`, so kinda expected.
For the second test, I opened firefox searched for notifications test and got cleverpush.com/en/test-notifications. dbus-monitor confirms this one sends image-data, but still got only qube icon. And a log entry on the server part: `Ignoring unusablke image from guest: Width or height too large`. It seems the icon is 256x256 here (BTW, that error could include that info...).

For the size, either the limit should be higher, or the client should scale it down to allowed size. The current limit seems to be 255, so almost... Maybe dbus notification spec includes suggested limits?

As for the `image-path`, if it's function is equivalent to `image-data`, maybe client could read the file and send as `image-data`?

**Comment by oshogbo** [comment]
> I generally like it, but I have questions: how does it look when the icon has some transparency? I wonder if this remains readable with colorful icons - I don't see that many colorful icons in notifications in general, though.

It shouldn't have any problems with transparency. I tested it, for example, with this:
2026-08-19-163425_531x129_scrot

> It doesn't appear to work :(

I haven't tested it with qubes, just directly on Debian. Let me try to address these problems.

**Comment by andrewdavidwong** [comment]
Suggestion: If you intend for this pull request to resolve the associated issue and would like for it to be [linked to the issue automatically](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue), you can put `Closes QubesOS/qubes-issues#XXXX` in the PR message (where `XXXX` is the issue number). `Closes` is one of several [special keywords](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests) that GitHub will recognize. If and when the PR is merged, GitHub will automatically close the associated issue.

**Comment by marmarek** [comment]
nm-applet now looks to work: https://openqa.qubes-os.org/tests/194312#step/startup/8 (click the image icon to see colors) :)

File: src/badge.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+    draw_cube(
+        &mut img,
+        edge * CUBE_MAIN_X,
+        edge * CUBE_MAIN_Y,
+        r,
+        colour,
+        1.0,
+    );
+    downsample(&img, SS)
+}
```
Comment (marmarek) [inline src/badge.rs]: The overlay was supposed to be just qube icon, which you can load from the icon theme based on `default_icon` of NotificationEmitter. You don't need to draw it manually...

**Review by marmarek — CHANGES_REQUESTED** [review]

File: src/badge.rs (hunk 1 above)
Comment (oshogbo) [reply src/badge.rs]: I somehow assumed that we wanted to support any color of the cube, not just the ones defined by the icons.

I’ll fix it then - to use the icons.

File: src/icon.rs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+    }
+    let file = format!("{name}.png");
+    let mut best: Option<((u8, u32), PathBuf)> = None;
+    for root in hicolor_roots() {
+        let size_dirs = match std::fs::read_dir(root) {
+            Ok(size_dirs) => size_dirs,
+            Err(_) => continue,
+        };
+        for size_dir in size_dirs.flatten() {
+            let size = match parse_icon_size(&size_dir.file_name()) {
```
Comment (marmarek) [inline src/icon.rs:92]: Shouldn't this check for file extension? It currently handles only png, but some icons are in svg...

**Comment by marmarek** [comment]
It is better now :)

Few more observations:
- If an application sends a large icon as `image-data`, the notification has noticeable delay, and at this time notification-proxy-client uses a lot of CPU. With 1024x1024 icon I got about 3s delay (on admittedly quite slow system). And then, as expected, dom0 rejects the icon anyway...

- After few more tests, it looks like lower right corner is unfortunate, as many icons use this area for various icon variants already (network manager for connection status (red cross if not disconnected), nextcloud for connection status, etc). So, maybe upper left corner instead?

File: src/icon.rs (hunk 2 above)
Comment (oshogbo) [reply src/icon.rs]: > let file = format!("{name}.png");

The lookup only asks for `<name>.png`.

**Comment by oshogbo** [comment]
I have moved the mark to the top-left. I'm not sure if it looks best there. Maybe bottom-left?                                                                                                      
        
The client now resizes images before sending them to dom0. I also optimized the client's parsing of the hints.
On my machine a 1024x1024 icon went from ~350 ms to ~67 ms.

**Comment by marmarek** [comment]
```
22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build] error[E0583]: file not found for module `hints`
22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build]   --> src/lib.rs:26:1
22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build]    |
22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build] 26 | mod hints;
22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build]    | ^^^^^^^^^^
22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build]    |
22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build]    = help: to create the module `hints`, create file "src/hints.rs" or "src/hints/mod.rs"
22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build]    = note: if there is a `mod hints` elsewhere in the crate already, import it with `use crate::...` instead
```
forgot to git add a file?

**Comment by oshogbo** [comment]
> ```
> 22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build] error[E0583]: file not found for module `hints`
> 22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build]   --> src/lib.rs:26:1
> 22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build]    |
> 22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build] 26 | mod hints;
> 22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build]    | ^^^^^^^^^^
> 22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build]    |
> 22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build]    = help: to create the module `hints`, create file "src/hints.rs" or "src/hints/mod.rs"
> 22:35:35 [qb.build_rpm.notification-proxy.vm-fc44.build]    = note: if there is a `mod hints` elsewhere in the crate already, import it with `use crate::...` instead
> ```
> 
> forgot to git add a file?

🤦‍♂️
Sorry for that, fixed.

**Comment by marmarek** [comment]
You're right, it doesn't look great...
https://openqa.qubes-os.org/tests/196375#step/startup/10
So, maybe indeed left bottom?

**Comment by oshogbo** [comment]
I have one more idea: maybe instead of moving the cube icon around, we could resize it and keep it in the bottom-right corner.

2026-09-21-164437_525x116_scrot
2026-09-21-164448_518x117_scrot
2026-09-21-164627_546x134_scrot
2026-09-21-164635_529x131_scrot

**Comment by marmarek** [comment]
Interesting idea, maybe? I'm a bit worried the scaled icon may be too small, but on the images above it looks okay. The benefit of this approach is it's easier to distinguish the icons when they don't fully overlay with the other.

**Comment by oshogbo** [comment]
I’ve pushed the code with a 75% reduction.

**Comment by marmarek** [comment]
I like it :)

**Review by marmarek — APPROVED** [review]
